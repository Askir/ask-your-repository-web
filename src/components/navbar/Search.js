// @flow
/* eslint-disable jsx-a11y/no-autofocus */
import React, { Component, Fragment } from 'react';
import onClickOutside from 'react-onclickoutside';
import { connect } from 'react-redux';
import { MdSearch, MdClose } from 'react-icons/md';
import classNames from 'classnames';
import ImageService from '../../services/ImageService';
import { startPresentation } from '../../state/presentation/presentation.actionCreators';
import type { AppState } from '../../state/AppState';
import type { Team } from '../../models/Team';
import './Search.scss';

type Props = {
  dispatch: Function,
  activeTeam: ?Team,
};

type State = {
  search: string,
  isSelected: boolean,
};

class Search extends Component<Props, State> {
  constructor(props: Props) {
    super(props);

    this.state = {
      isSelected: false,
      search: '',
    };
  }

  handleClickOutside = () => {
    this.setState({ isSelected: false });
  };

  handleSelect = () => {
    this.setState({ isSelected: true });
  };

  handleClose = () => {
    this.setState({ isSelected: false });
  };

  handleChange = (e: SyntheticInputEvent<HTMLInputElement>) => {
    const { value } = e.target;
    this.setState({ search: value });
  };

  handleSubmit = async (e: SyntheticEvent<HTMLFormElement>) => {
    e.preventDefault();

    const { activeTeam } = this.props;
    if (!activeTeam) return;

    const { search } = this.state;
    const images = await ImageService.list({
      teamId: activeTeam.id,
      search,
    });

    const filteredImages = images
      .filter(image => image.score > 0)
      .slice(0, 4);

    this.props.dispatch(startPresentation(filteredImages));
  };

  render() {
    const { isSelected, search } = this.state;
    const className = classNames('Search', {
      'Search--active': isSelected,
    });

    return (
      <form className={className} onSubmit={this.handleSubmit}>
        {isSelected ? (
          <Fragment>
            <MdSearch className="Search__input__icon Search__input__icon--left" />
            <input
              autoFocus
              type="text"
              value={search}
              className="Search__input"
              onChange={this.handleChange}
            />
            <button
              type="button"
              onClick={this.handleClose}
              className="Search__input__close"
            >
              <MdClose />
            </button>
          </Fragment>
        ) : (
          <button
            type="button"
            onClick={this.handleSelect}
            className="Search__preview"
          >
            <MdSearch className="Search__preview__icon" />
            <span className="Search__preview__text">
              Suchen
            </span>
          </button>
        )}
      </form>
    );
  }
}

const mapStateToProps = (state: AppState) => ({
  activeTeam: state.activeTeam,
});

export default connect(mapStateToProps)(onClickOutside(Search));
