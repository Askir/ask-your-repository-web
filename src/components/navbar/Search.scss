@import '../../style/breakpoints';

.Search {
  --margin-vertical: 10px;

  --height-mobile: var(--searchbar-height);
  --width-mobile: 100%;

  --width-desktop: 330px;
  --height-desktop: calc(var(--navbar-height) - 2 * var(--margin-vertical));

  position: relative;
  height: var(--height-mobile);
  width: var(--width-mobile);

  background-color: var(--accent-color-blue);
  box-shadow: 0 0 10px 5px rgba(0, 0, 0, .2);
  border-radius: 4px;

  font-size: 1.5rem;

  transition: all .2s;

  &--active {
    width: var(--width-mobile);
    height: var(--height-mobile);
  }

  @media (min-width: $sm-to-md) {
    position: relative;
    width: var(--width-desktop);
    height: var(--height-desktop);
    margin: var(--margin-vertical) 0;
  }

  &__preview, &__input {
    width: 100%;
    height: 100%;
    display: flex;
    align-items: center;
    justify-content: center;
    background-color: transparent;
  }

  &__preview {
    cursor: pointer;
    width: 100%;
    color: rgba(255, 255, 255, .5);
    outline: none;

    &__text {
      margin-left: 10px;
    }
  }

  &__input {
    height: 100%;
    padding: 20px;
    padding-left: 60px;
    outline: 0;
    position: relative;

    &__icon, &__close {
      position: absolute;
      top: 50%;
      transform: translateY(-50%);
      z-index: 60;
    }

    &__icon {
      left: 20px;
    }

    &__close {
      display: flex;
      align-items: center;
      justify-content: center;
      right: 20px;
      background: transparent;
    }
  }
}
