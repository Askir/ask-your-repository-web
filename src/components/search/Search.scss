.Search {
  z-index: 50;
  position: fixed;
  top: 20px;
  left: auto;
  right: auto;

  font-size: 2rem;

  &__preview, &__input {
    width: 440px;
    height: 80px;

    display: flex;
    align-items: center;
    justify-content: center;
    background-color: rgba(38, 124, 221, 0.98);
    box-shadow: 0 0 10px 5px rgba(0, 0, 0, .2);
  }

  &__preview {
    cursor: pointer;
    color: rgba(255, 255, 255, .5);

    &__icon {
      margin-right: 10px;
    }
  }

  &__input {
    padding: 20px;
    padding-left: 70px;
    outline: 0;
    position: relative;

    &__icon {
      top: 50%;
      transform: translateY(-50%);
      left: 20px;
      z-index: 60;
      position: absolute;
    }
  }
}
