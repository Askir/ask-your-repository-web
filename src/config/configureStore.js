// @flow
import { applyMiddleware, createStore } from 'redux';
import thunk from 'redux-thunk';
import rootReducer from '../state/rootReducer';

function configureStore() {
  let middleware = [thunk];

  if (process.env.NODE_ENV !== 'production') {
    const { logger } = require('redux-logger');
    middleware = [...middleware, logger];
  }

  return createStore(
    rootReducer,
    applyMiddleware(...middleware),
  );
}

export default configureStore;
