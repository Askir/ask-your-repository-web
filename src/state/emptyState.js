// @flow
import type { AppState } from './AppState';

const emptyState: AppState = {
  image: null,
  images: [],
};

export default emptyState;
