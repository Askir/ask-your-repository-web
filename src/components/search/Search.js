// @flow
import React, { Component, Fragment } from 'react';
import { connect } from 'react-redux';
import { push } from 'connected-react-router';
import { MdSearch } from 'react-icons/md';
import api from '../../config/api';
import { startPresentation } from '../../state/presentation/presentation.actionCreators';
import './Search.scss';

type Props = {
  dispatch: Function,
};

type State = {
  search: string,
  isSelected: boolean,
};

class Search extends Component<Props, State> {
  constructor(props: Props) {
    super(props);

    this.state = {
      isSelected: false,
      search: '',
    };
  }

  handleSelect = () => {
    this.setState({ isSelected: true });
  };

  handleChange = (e: SyntheticInputEvent<HTMLInputElement>) => {
    const { value } = e.target;
    this.setState({ search: value });
  };

  handleSubmit = async (e: SyntheticEvent<HTMLFormElement>) => {
    e.preventDefault();
    const { search } = this.state;
    const { data } = await api.get('/images', { params: { search } });

    this.props.dispatch(startPresentation(data.images.slice(0, 4)));
    this.props.dispatch(push('/presentation'));
  };

  render() {
    const { isSelected, search } = this.state;

    return (
      <form className="Search" onSubmit={this.handleSubmit}>
        {isSelected ? (
          <Fragment>
            <MdSearch className="Search__input__icon" />
            <input
              autoFocus
              type="text"
              value={search}
              className="Search__input"
              onChange={this.handleChange}
            />
          </Fragment>
        ) : (
          <button
            type="button"
            onClick={this.handleSelect}
            className="Search__preview"
          >
            <MdSearch className="Search__preview__icon" />
            Suchen
          </button>
        )}
      </form>
    );
  }
}

export default connect()(Search);
