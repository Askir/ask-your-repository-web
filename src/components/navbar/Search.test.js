// @flow
import React from 'react';
import toJson from 'enzyme-to-json';
import { shallow } from 'enzyme';
import configureStore from 'redux-mock-store';
import emptyState from '../../state/emptyState';
import Search from './Search';

const mockStore = configureStore();


describe('<Search />', () => {
  let wrapper;
  let store;

  beforeEach(() => {
    store = mockStore(emptyState);
    wrapper = shallow(<Search store={store} />).dive();
  });

  it('renders correctly', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
