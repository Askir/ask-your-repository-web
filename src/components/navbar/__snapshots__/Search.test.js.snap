// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Search /> renders correctly 1`] = `
<form
  className="Search"
  onSubmit={[Function]}
>
  <button
    className="Search__preview"
    onClick={[Function]}
    type="button"
  >
    <MdSearch
      className="Search__preview__icon"
    />
    Suchen
  </button>
</form>
`;
